File: leetcode/p00234/p00234.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next =next

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head.next:
            return True
        
        slow = head
        fast = head

        while fast and fast.next:
            fast = fast.next.next
            if not fast:
                tmp = slow.next
                slow.next = None
                slow = tmp
            else:
                slow = slow.next
        
        previous = slow
        current = previous.next
        previous.next = None

        while current:
            tmp = current.next
            current.next = previous
            previous = current
            current = tmp
        
        tail = previous
        while head and tail:
            if head.val != tail.val:
                return False
            
            head = head.next
            tail = tail.next
        
        return True

if __name__ == '__main__':
    s = Solution()

    list = [ListNode(1), ListNode(2), ListNode(2), ListNode(1)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    assert s.isPalindrome(list[0]) == True

    list = [ListNode(1), ListNode(2)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    assert s.isPalindrome(list[0]) == False

File: leetcode/p00234/p00234.go
package p00234

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time Complexity: O(n)
// Space Complexity: O(1)
func isPalindrome(head *ListNode) bool {
	// step.1 find the middle of linked list
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	// 當 linked list 長度是奇數時, 需要往後位移一個 node 才是要判斷回文的起點(終點)
	if fast != nil {
		slow = slow.Next
	}

	// step.2 reverse the half of linked list
	var last, current, next *ListNode
	current = slow
	for current != nil {
		next = current.Next
		current.Next = last
		last, current = current, next
	}

	// step.3 check palindrome
	left, right := head, last
	for right != nil {
		if left.Val != right.Val {
			return false
		}
		left = left.Next
		right = right.Next
	}
	return true
}
